### ISSUE TITLE
async arrow functions generate invalid scoping

### ISSUE THREAD

**Issue by TimWolla** [babel/babel#6908]
Choose one: bug report

<strike>Double nested async functions generate code that has incorrect variable scoping if destructuring is used for the parameters.</strike> (incorrect, see comment)

Async arrow functions generate invalid scoping.

### Input Code

```js
const regeneratorRuntime = require('regenerator-runtime')

async function foo() {
	const promises = [ [ 1 ], [ 2 ], [ 3 ] ].map(async ([ number ]) => {
		return new Promise(function (resolve, reject) {
			setTimeout(function () {
				resolve(number)
			}, 1)
		});
	})

	return Promise.all(promises)
}

foo().then(function (number) {
	console.log(number)
});
```

### Babel/Babylon Configuration (.babelrc, package.json, cli command)

```js
{ "presets": [ [ "env"
               , { "targets": { "browsers": [ "last 2 chrome versions"
                                            , "last 2 chromeandroid versions"
                                            , "firefox esr"
                                            , "last 2 firefox versions"
                                            , "edge >= 15"
                                            , "safari >= 11"
                                            ]
                              }
                 , "debug": true
                 }
               ]
             ]
}
```

### Expected Behavior

The code should print `[ 1, 2, 3 ]`, like node does:

```
root@222243f2e5a6:/pwd# node test.js 
[ 1, 2, 3 ]
```

### Current Behavior

The code instead prints `[ 3, 3, 3 ]`, because `number` is incorrectly scoped:

```
root@222243f2e5a6:/pwd# babel test.js -o test.babel.js ; node test.babel.js 
babel-preset-env: `DEBUG` option

Using targets:
{
  "chrome": "61",
  "edge": "15",
  "firefox": "52",
  "safari": "11"
}

Modules transform: commonjs

Using plugins:
  transform-es2015-destructuring {"edge":"15","firefox":"52"}
  transform-es2015-for-of {"firefox":"52"}
  transform-es2015-function-name {"edge":"15","firefox":"52"}
  transform-es2015-literals {"firefox":"52"}
  transform-es2015-parameters {"firefox":"52"}
  transform-regenerator {"firefox":"52"}
[ 3, 3, 3 ]
```

### Possible Solution

The declaration of `number` should not be hoisted to the outer function.

### Context

<strike>I experienced the error while trying to `map` an `async` function across the return value of `Object.entries` (that's why I wanted to destructure the parameter of the closure).</strike> (no longer relevant, see comment)

### Your Environment

| software         | version(s)
| ---------------- | -------
| Babel            |  6.26.0 (babel-core 6.26.0)
| Babylon          | 
| node             | v9.2.0
| npm              | 5.5.1
| Operating System | Official node.js Docker container

**Comment by TimWolla** [comment]
Some further test cases. It is broken even with a simple `var`:

```
const regeneratorRuntime = require('regenerator-runtime')

async function foo() {
	const promises = [ 1, 2, 3 ].map(async (number) => {
		var tmp = number
		return new Promise(function (resolve, reject) {
			setTimeout(function () {
				resolve(tmp)
			}, 1)
		})
	})

	return Promise.all(promises)
}

foo().then(function (number) {
	console.log(number)
});
```

It *works fine* if I replace the arrow function with a regular function, the variable is correctly scoped in the generated JavaScript:

```
const regeneratorRuntime = require('regenerator-runtime')

async function foo() {
	const promises = [ 1, 2, 3 ].map(async function (number) {
		var tmp = number
		return new Promise(function (resolve, reject) {
			setTimeout(function () {
				resolve(tmp)
			}, 1)
		})
	})

	return Promise.all(promises)
}

foo().then(function (number) {
	console.log(number)
});
```

**Comment by loganfsmyth** [comment]
Would you be able to post a snippet of what the output code looks like?

**Comment by TimWolla** [comment]
@loganfsmyth Sure:

```js
// root@cebff537eeed:/pwd# cat test.js 
const regeneratorRuntime = require('regenerator-runtime')

async function foo() {
	const promises = [ 1, 2, 3 ].map(async (number) => {
		var tmp = number
		return new Promise(function (resolve, reject) {
			setTimeout(function () {
				resolve(tmp)
			}, 1)
		})
	})

	return Promise.all(promises)
}

foo().then(function (number) {
	console.log(number)
});
// root@cebff537eeed:/pwd# cat test.babel.js 
'use strict';

const regeneratorRuntime = require('regenerator-runtime');

function foo() {
	var tmp, promises;
	return regeneratorRuntime.async(function foo$(_context2) {
		while (1) switch (_context2.prev = _context2.next) {
			case 0:
				promises = [1, 2, 3].map(number => {
					return regeneratorRuntime.async(function _callee$(_context) {
						while (1) switch (_context.prev = _context.next) {
							case 0:
								tmp = number;
								return _context.abrupt('return', new Promise(function (resolve, reject) {
									setTimeout(function () {
										resolve(tmp);
									}, 1);
								}));

							case 2:
							case 'end':
								return _context.stop();
						}
					}, null, this);
				});
				return _context2.abrupt('return', Promise.all(promises));

			case 2:
			case 'end':
				return _context2.stop();
		}
	}, null, this);
}

foo().then(function (number) {
	console.log(number);
});
```

**Comment by loganfsmyth** [comment]
Could you try it with `include: ['transform-es2015-arrow-functions']` in your options for `babel-preset-env` as a sibling of `debug` and `targets`? That might be enough for a quick fix for you.

**Comment by TimWolla** [comment]
> That might be enough for a quick fix for you.

Yes, that fixed the issue. Thank you.

**Comment by loganfsmyth** [comment]
Great. Definitely agree this is a bug. Either we'll need to update regenerator to handle this, or require that the arrow transform run whenever regenerator is in use. At least historically, regenerator has assumed that the content inside an async function or generator has been simple ES5, but I'm not sure if that's still the goal or if they are open to updating it for more ES6 cases.

**Comment by TimWolla** [comment]
> or require that the arrow transform run whenever regenerator is in use

I suggest that one. Both arrow functions as well as generators are supported in browsers since forever (so generally you would require the arrow function transform for browser support already).

In fact I wonder why regenerator is pulled in for Firefox 52 (that's the ESR) if MDN indicates full generator support as of Firefox 38? Is it disabled with a feature flag in the ESR? Or is it a bug in the env preset?

**Comment by TimWolla** [comment]
@loganfsmyth It was easier than expected: The issue should be fixed as of facebook/regenerator@682a3fcac9389791ee63914a7ad0cf233536a8b1.

I don't know whether you have to act on this (e.g. by updating some `package.json`), that's why I wanted to make you aware of it.
